1.Thrift store

In [65]:
using JuMP, Cbc

m = Model(solver=CbcSolver())

#penny
@variable(m, p >= 0, Int)
#nickel
@variable(m, n >= 0, Int)
#dime
@variable(m, d >= 0, Int)
#quarter
@variable(m, q >= 0, Int)

@constraint(m, p + 5n + 10d + 25q == 99)
@objective(m, Min, 2.5p + 5n + 2.268d + 5.670q)

status = solve(m)
println("Penny: ", getvalue(p))
println("Nickel: ", getvalue(n))
println("Dime: ", getvalue(d))
println("Quarter", getvalue(q))
println("Total weight:", getobjectivevalue(m))

Penny: 4.0
Nickel: 0.0
Dime: 7.000000000000001
Quarter1.0
Total weight:31.546


According to the Cbc solver, we should have 4 pennies, 7 dimes and 1 quarter. And the total weight is 31.546 grams.




2.Checked luggage 

In [66]:
using JuMP, Gurobi

weights = [5,6,7,6,4,6,7,3,8,5]
volume = [2,4,5,3,3,2,3,1,2,4]

m = Model(solver=GurobiSolver(OutputFlag=0))

#souvenirs: 1 taken, 0 not taken
@variable(m, s[1:10], Bin)
#weight constraint
@constraint(m, sum(s.*weights) <= 30)
#volume constraint
@constraint(m, sum(s.*volume) <= 15)
@objective(m, Max, sum(s))

status = solve(m)
println(getvalue(s))
println("Number of souvenirs to bring: ", sum(getvalue(s)))

Academic license - for non-commercial use only
[1.0, -0.0, -0.0, 1.0, 1.0, 1.0, -0.0, 1.0, 0.0, 1.0]
Number of souvenirs to bring: 6.0


According to the Gurobi solver, we should take Souvenir 1, 4, 5, 6, 8 and 10.




3.Combat computers

In [67]:
using JuMP, Gurobi

m = Model(solver=GurobiSolver(OutputFlag=0))
#binary variable indicating whather we are using a production plant
@variable(m, z[1:4], Bin)
#how many computers are produced by each production plant
@variable(m, x[1:4] >= 0, Int)
@constraint(m, sum(x) <= 20000)
@constraint(m, x[1] <= 10000)
@constraint(m, x[2] <= 8000)
@constraint(m, x[3] <= 9000)
@constraint(m, x[4] <= 6000)
@constraint(m, x[1] <= 10000*z[1])
@constraint(m, x[2] <= 8000*z[2])
@constraint(m, x[3] <= 9000*z[3])
@constraint(m, x[4] <= 6000*z[4])
@objective(m, Max, sum(x)*3500 - (x[1]*1000+x[2]*1700+x[3]*2300+x[4]*2900) - 
    (z[1]*9000000+z[2]*5000000+z[3]*3000000+z[4]*1000000))
status = solve(m)

println(getvalue(x))

Academic license - for non-commercial use only
[10000.0, 8000.0, -0.0, 2000.0]


According to the solver, we should use plant 1 to produce 10000 computers, plant 2 to produce 8000 computers, we don't use plant 3, and use plant 4 to produce 2000 computers.




4.ABC Investments

In [68]:
using JuMP, Cbc

expected_return = [0.13 0.09 0.17 0.1 0.22 0.12]
minInvest = [3 2 9 5 12 4]
maxInvest = [27 12 35 15 46 18]

m = Model(solver=CbcSolver())
@variable(m, z[1:6], Bin)
@variable(m, x[1:6] >= 0)
@constraint(m, minConstr[i in 1:6], x[i] >= minInvest[i].*z[i])
@constraint(m, maxConstr[i in 1:6], x[i] <= maxInvest[i].*z[i])
@constraint(m, x[5] <= x[2]+x[4]+x[6])
@constraint(m, x[6] >= z[3]*minInvest[6])
@constraint(m, x[6] <= z[3]*maxInvest[6])
@constraint(m, sum(x[1:6]) <= 80)

@objective(m, Max, sum(x[i] * expected_return[i] for i in 1:6))

status = solve(m)
println(getvalue(x))
println("Total profit: ", getobjectivevalue(m))

[0.0, 0.0, 35.0, 5.0, 22.5, 17.5]
Total profit: 13.500000000000004


Therefore, we shouldn't invest anything on option 1 and option 2, we should invest 35 million on option 3, 5 million on option 4, 22.5 million on option 5 and 17.5 million to option 6. This will give us a total profit of 13.500000000000004 million.